File: src/295_find_median_from_data_stream/FindMedianFromDataStream.java
import java.util.*;

public class Solution {
    PriorityQueue<Integer> small;
    PriorityQueue<Integer> large;

    public Solution() {
        small = new PriorityQueue<>(Collections.reverseOrder());
        large = new PriorityQueue<>();
    }

    public void addNum(int num) {
        small.offer(num);
        large.offer(small.poll());
        if (large.size() > small.size()) {
            small.offer(large.poll());
        }
    }

    public Double findMedian() {
        if (small.size() > large.size()) {
            return (double)small.peek();
        }

        return (double)(small.peek() + large.peek()) / 2;
    }
}

public class FindMedianFromDataStream {
    public static void main(String[] args) {
        Solution sol = new Solution();
        sol.addNum(1);
        sol.addNum(2);
        System.out.println(sol.findMedian() + " | " + 1.5);
        sol.addNum(3);
        System.out.println(sol.findMedian() + " | " + 2.0);
        Solution sol2 = new Solution();
        for (int n : new int[]{5, 3, 8, 9, 2}) sol2.addNum(n);
        System.out.println(sol2.findMedian() + " | " + 5);
        Solution sol3 = new Solution();
        for (int n : new int[]{10, 20, 30, 40}) sol3.addNum(n);
        System.out.println(sol3.findMedian() + " | " + 25.0);
        Solution sol4 = new Solution();
        sol4.addNum(7);
        System.out.println(sol4.findMedian() + " | " + 7);
    }
}

File: src/295_find_median_from_data_stream/find_median_from_data_stream.cpp
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

class Solution {
    priority_queue<int> small;                             // max-heap
    priority_queue<int, vector<int>, greater<int>> large;  // min-heap
   public:
    Solution() {}
    void addNum(int num) {
        small.push(num);
        large.push(small.top());
        small.pop();
        if (large.size() > small.size()) {
            small.push(large.top());
            large.pop();
        }
    }
    double findMedian() {
        if (small.size() > large.size()) {
            return small.top();
        }
        return (small.top() + large.top()) / 2.0;
    }
};

// Test scaffolding
int main() {
    Solution sol;
    sol.addNum(1);
    sol.addNum(2);
    cout << sol.findMedian() << " | " << 1.5 << endl;
    sol.addNum(3);
    cout << sol.findMedian() << " | " << 2.0 << endl;
    Solution sol2;
    for (int n : vector<int>{5, 3, 8, 9, 2}) sol2.addNum(n);
    cout << sol2.findMedian() << " | " << 5 << endl;
    Solution sol3;
    for (int n : vector<int>{10, 20, 30, 40}) sol3.addNum(n);
    cout << sol3.findMedian() << " | " << 25.0 << endl;
    Solution sol4;
    sol4.addNum(7);
    cout << sol4.findMedian() << " | " << 7 << endl;
    return 0;
}
